### ISSUE TITLE
SameDiff CUDA performance analysis

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#8560]
densenet: https://gist.github.com/AlexDBlack/2503742ca13c31358ceb53f14abb9a86
squeezenet: https://gist.github.com/AlexDBlack/5341036f60c4090ca8e2439c6d79d104
mobilenet_v1: https://gist.github.com/AlexDBlack/bf63b3035e78770e85921a9d73fa7089

Obviously we ignore batchnorm and conv2d etc, as SameDiff is running without cuDNN (and TF with). SameDiff cuDNN is in progress.

Hotspots:
* concat (densenet, squeezenet, )
* reduce mean (along dimension? densenet)
* argmax (along dimensions? densenet)
* biasadd (squeezenet, mobilenetv1)
* stack (squeezenet)
* reshape (squeezenet, mobilenetv1), squeeze (mobilenetv1)
* strided_slice (squeezenet)
* softmax (along dim? squeezenet, mobilenetv1)
* relu / relu6 (squeezenet, mobilenetv1)

**Comment by AlexDBlack** [comment]
```
densenet block-2/denseblock-2-13/concat
in0: [4,32,14,14,672,131712,9408,672,1,8192,1,99]
in1: [4,32,14,14,32,6272,448,32,1,8192,1,99]
in2: [0,131072,1,99] - value [3]
bargs: [true]
out0: [4,32,14,14,704,137984,9856,704,1,8192,1,99]
```
Concat on last axis of (standard strides) C order can be a memcpy - i.e., concatenate buffers

```
densenet global_pool (reduce_mean)
x: [4,32,7,7,1024,50176,7168,1024,1,8192,1,99]
z: [2,32,1024,1024,1,8192,1,99]
dimensions: [1,2]
```

```
densenet ArgMax
in0: [2,32,1001,1001,1,8192,1,99]
in1: [0,131072,1,99] - value 1
out0: [1,32,1,262144,1,99]
```

```
squeezenet flatten/Reshape/shape (stack op)
in0: [0,131072,1,99] - value 32
in1: [0,131072,1,99] - value -1
out0: [1,2,1,131072,1,99] - value [32,-1]
```

```
squeezenet flatten/Reshape (reshape)
in0: [4,32,1,1,1001,1001,1001,1001,1,8192,1,99]
in1: [1,2,1,131072,1,99] - value [32, -1]
out0: [2,32,1001,1001,1,8192,1,99]
```

```
squeezenet flatten/strided_slice
in0: [1,4,1,131072,1,99] - values [        32,         1,         1,      1001]
in1: [1,1,1,131072,1,99] - values [0]
in2: [1,1,1,131072,1,99] - values [1]
in3: [1,1,1,131072,1,99] - values [1]
in4: [0,131072,1,99] - values [32]
iargs [0,0,0,0,1]
```

```
squeezenet softmax
in0: [2,32,1001,1001,1,8192,1,99]
out0: [2,32,1001,1001,1,8192,1,99]
```

```
squeezenet conv1/Relu
x: [4,32,112,112,96,1204224,10752,96,1,8192,1,99]
z: [4,32,112,112,96,1204224,10752,96,1,8192,1,99]
scalar: 0
```

**Comment by AlexDBlack** [comment]
ReLU time - TF time is physically impossible on that card, which suggests it's async execution
https://gist.github.com/AlexDBlack/65c24030b7a8d3d71168f346b6188bd9
Seeing about 61% of theoretical memory bandwidth for GTX970 (120 vs. 196 GB/s spec) in both synthetic test and SameDiff op execution.
TF being 8.65x faster is simply not possible.

Edit: for a fair / non-biased comparison, I'll need to find a way to enable synchronous CUDA execution for TensorFlow.

**Comment by AlexDBlack** [comment]
No luck so far in finding a way to force synchronous execution (for TF or CUDA generally)
The only thing I've been able to find is references to "CUDA_LAUNCH_BLOCKING" - which doesn't seem to make a difference...
(one reference also said it didn't impact CUDA streams API)
I have implemented aggregation of TF sub-ops for CUDA, as well as filtering of TF's op names, 

At any rate, aggregated / cleaned up CUDA results are here:
I'm still not sure how much these results can be trusted - I suspect there's a mix of "real" times (via cuda profiling/timing API) and "async" times.

densenet: https://gist.github.com/AlexDBlack/4af90f3351981b3f0b655e584401d1e0
squeezenet: https://gist.github.com/AlexDBlack/271201d83d2f722d12d1abc196a2b91d
nasnet_mobile: https://gist.github.com/AlexDBlack/8fd3baa479600e606bd356a95017bab1
mobilenet_v1: https://gist.github.com/AlexDBlack/732a33a96ddb80795dad4bed5cb75211

**Comment by saudet** [comment]
[`cudaDeviceSynchronize()`](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g10e20b05a95f638a4071a655503df25d) is what you're looking for I believe?

**Comment by AlexDBlack** [comment]
@saudet Yes, but I don't believe I can (in any easy/practical sense) insert that into tensorflow after each operation's execution... that's what I really need here to ensure the profiler timing is "real" not just counting launch time.

**Comment by AlexDBlack** [comment]
SameDiff CUDA vs. CPU
densenet: https://gist.github.com/AlexDBlack/e57c0bff9e551c87050318b0a1f1d90a
squeezenet: https://gist.github.com/AlexDBlack/fc5dc5303e8d8f45a66776513f2fb698
nasnet_mobile: https://gist.github.com/AlexDBlack/3dfbbc93947bb4462facb0c53cd01eb3
inception_resnet_v2: https://gist.github.com/AlexDBlack/0d444f6e985f4126da0841952d6aa60e

**Comment by AlexDBlack** [comment]
Note above tests were ubuntu 18.04, CUDA 10.2, gtx 970

To reproduce: https://github.com/KonduitAI/dl4j-dev-tools/pull/35
Links to models can be found here: https://github.com/KonduitAI/dl4j-dev-tools/blob/master/import-tests/profiling/zoo_model_profiling.py
(download + extract)

Shapes from a few of the above tests:
```
densenet batch_normalization/FusedBatchNorm:
inputs:
[4,32,112,112,64,802816,7168,64,1,8192,1,99]
[1,64,1,8192,1,99]
[1,64,1,8192,1,99]
[1,64,1,8192,1,99]
[1,64,1,8192,1,99]
outputs:
[4,32,112,112,64,802816,7168,64,1,8192,1,99]
targs: [1.0009999641624745E-5]
iargs: [1,1,3]
```

```
densenet block-2/denseblock-2-0/batch_normalization/FusedBatchNorm
inputs:
[4,32,14,14,256,50176,3584,256,1,8192,1,99]
[1,256,1,8192,1,99]
[1,256,1,8192,1,99]
[1,256,1,8192,1,99]
[1,256,1,8192,1,99]
outputs:
[4,32,14,14,256,50176,3584,256,1,8192,1,99]
targs: 1.0009999641624745E-5
iargs: [1,1,3]
```

```
densenet global_pool (reduce_mean)
x = [4,32,7,7,1024,50176,7168,1024,1,8192,1,99]
z = [2,32,1024,1024,1,8192,1,99]
dimensions = [1, 2]
```

```
squeezenet cell_11/cell_output/concat
inputs:
[4,32,7,7,176,8624,1232,176,1,8192,1,99]
[4,32,7,7,176,8624,1232,176,1,8192,1,99]
[4,32,7,7,176,8624,1232,176,1,8192,1,99]
[4,32,7,7,176,8624,1232,176,1,8192,1,99]
[4,32,7,7,176,8624,1232,176,1,8192,1,99]
[4,32,7,7,176,8624,1232,176,1,8192,1,99]
[0,131072,1,99] - value 3
outputs:
[4,32,7,7,1056,51744,7392,1056,1,8192,1,99]
bargs: [true]
```

**Comment by raver119** [comment]
Concat: https://github.com/KonduitAI/deeplearning4j/pull/148/files

However, for me (2060S)  it was 350-440 us before tweak, 200-250 us after tweak.

**Comment by AlexDBlack** [comment]
retesting concat on CUDA GTX970 after recent optimizations:

```
Densenet:                           Before      After
block-0/denseblock-0-5/concat       141225      129498
block-0/denseblock-0-4/concat       124021      110884
block-1/denseblock-1-11/concat      73232       71772
```

So, definitely faster, but not much faster...

```
block-0/denseblock-0-4/concat
inputs:
[4,32,56,56,192,602112,10752,192,1,8192,1,99]
[4,32,56,56,32,100352,1792,32,1,8192,1,99]
[0,131072,1,99] - value 3
```

```
block-0/denseblock-0-5/concat
[4,32,56,56,224,702464,12544,224,1,8192,1,99]
[4,32,56,56,32,100352,1792,32,1,8192,1,99]
[0,131072,1,99] - value 3
```

```
block-1/denseblock-1-11/concat
[4,32,28,28,480,376320,13440,480,1,8192,1,99]
[4,32,28,28,32,25088,896,32,1,8192,1,99]
[0,131072,1,99] - value 3
```

**Comment by AlexDBlack** [comment]
Testing concat on GTX970 (CUDA 10.2, ubuntu 18.04)
https://gist.github.com/AlexDBlack/a7371f5c3bf1b209a253214a8038a913
Earlier test by @raver119 (that should have been the same dimensionality as the first case), Windows 2080ti:
```
$ ./blasbuild/cuda/tests_cpu/layers_tests/runtests.exe --gtest_filter="PlaygroundTests.test_s_1"
Note: Google Test filter = PlaygroundTests.test_s_1
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from PlaygroundTests
[ RUN      ] PlaygroundTests.test_s_1

Time: 1419 us;
[       OK ] PlaygroundTests.test_s_1 (3528 ms)
[----------] 1 test from PlaygroundTests (3528 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (3528 ms total)
[  PASSED  ] 1 test.
```

Need to validate that result on other GPUs, but a GTX970 should be nowhere near that slow in comparison to a 2080ti...

**Comment by raver119** [comment]
First suspect here would be int64 arithmetics used for indexing these days

**Comment by agibsonccc** [comment]
Closing — the CUDA backend has been completely rewritten with new memory management, DSP execution engine, CUDA graph replay, and Triton compilation. If you still experience GPU-related issues with the upcoming release, please open a new issue.
